sonic-thermalctld/scripts/thermalctld

```diff
@@ -0,0 +1,521 @@
+    from sonic_daemon_base.task_base import ProcessTaskBase
+except ImportError, e:
+    raise ImportError(str(e) + " - required module not found")
+
+try:
+    from swsscommon import swsscommon
+except ImportError as e:
+    from tests import mock_swsscommon as swsscommon
+
+
+SYSLOG_IDENTIFIER = 'thermalctld'
+NOT_AVAILABLE = 'N/A'
+logger = Logger(SYSLOG_IDENTIFIER)
+
+
+# utility functions
+
+# try get information from platform API and return a default value if caught NotImplementedError
+def try_get(callback, default=NOT_AVAILABLE):
+    """
+    Handy function to invoke the callback and catch NotImplementedError
+    :param callback: Callback to be invoked
+    :param default: Default return value if exception occur
+    :return: Default return value if exception occur else return value of the callback
+    """
+    try:
+        ret = callback()
+    except NotImplementedError:
+        ret = None
+
+    return default if ret is None else ret
+
+
+def log_on_status_changed(normal_status, normal_log, abnormal_log):
+    """
+    Log when any status changed
+    :param normal_status: Expected status.
+    :param normal_log: Log string for expected status.
+    :param abnormal_log: Log string for unexpected status
+    :return:
+    """
+    if normal_status:
+        logger.log_notice(normal_log)
+    else:
+        logger.log_warning(abnormal_log)
+
+
+class FanStatus(object):
+    def __init__(self):
+        """
+        Constructor of FanStatus
+        """
+        self.presence = True
+        self.under_speed = False
+        self.over_speed = False
+        self.invalid_direction = False
+
+    def set_presence(self, presence):
+        """
+        Set and cache Fan presence status
+        :param presence: Fan presence status
+        :return: True if status changed else False
+        """
+        if presence == self.presence:
+            return False
+
+        self.presence = presence
+        return True
+
+    def _check_speed_value_available(self, speed, target_speed, tolerance, current_status):
+        if speed == NOT_AVAILABLE or target_speed == NOT_AVAILABLE or tolerance == NOT_AVAILABLE:
+            if current_status is True:
+                logger.log_warning('Fan speed or target_speed or tolerance become unavailable.')
+            return False
+        return True
+
+    def set_under_speed(self, speed, target_speed, tolerance):
+        """
+        Set and cache Fan under speed status
+        :param speed: Fan speed
+        :param target_speed: Fan target speed
+        :param tolerance: Threshold between Fan speed and target speed
+        :return: True if status changed else False
+        """
+        if not self._check_speed_value_available(speed, target_speed, tolerance, self.under_speed):
+            old_status = self.under_speed
+            self.under_speed = False
+            return old_status != self.under_speed
+
+        status = (target_speed - speed > tolerance)
```

Comment:
this is not the correct way to check over speed or under-speed, should be:    target_speed*(1-tolerance) < speed < target_speed*(1+tolerance)